### ISSUE TITLE
[cli] docker events 'hangs' / stops working

### ISSUE THREAD

**Issue by dreamcat4** [docker/docker#12023]
Hello. Sorry to have to report this!

Calling the `docker events` command --> `while read …` loop. Via a FIFO (unix socket). It is meant to prints to stdout a new line of infos, whenever a container starts.

My code is here:

https://github.com/dreamcat4/docker-images/blob/master/pipework/entrypoint.sh#L368

To begin with, it works as it should. If I start 1 container, the while loop is triggered 1 time. Soon after that, I may restart the container (to generate a 2nd start event). And `docker events` will print out another line, triggering my `while read …` loop correctly once again. That is great. Well done.

However if I go away and leave the docker containers alone for "a while" (such overnight, or just 1 hour, or perhaps less <  1hour). Then I come back and start a container. There is nothing received in the unix pipe. And it seems almost like docker events is hung, or not flush anymore it's stdout. Or else not printed a newline `\n` for to trigger the next `read line`.

If I `kill $PID` of `docker events` command's process id. On a regular kill (which is a `SIGTERM`). Then sometimes doing that will prompt the hung `docker events` process to finally spit out whatever missing line(s) e.g. like a flush. Presumably during it's cleanup, and then quit. Else sometimes it may not print anything and just dies / go away. Sorry that sounds very vague. But it is difficult to understand why this inconsistent behaviour.

My main point is that I really need some help here. Because I am not sure how to progress from this point, and debug further the problem going forwards. It is an important bug for me to solve, one way or another (and without dropping the functionality!).

Is there any new tests I can make to catch it red-handed? What else can I do? Perhaps write some test example script (not my 'real' code), with a loop, to print out 'SUCCESS' / 'FAIL' record. And a constantly re-starting container in another loop? E.g. with some date / time elapsed recorded?

How else to get more infos about the bug? What about `strace` command on `docker events`? Or how about if we try to attach gdb and get a live stacktrace when it stops responding? What can be honed in on first of all?

```sh
$ docker info
Containers: 4
Images: 70
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 78
Execution Driver: native-0.2
Kernel Version: 3.16.0-33-generic
Operating System: Ubuntu 14.10
CPUs: 2
Total Memory: 3.792 GiB
```

**Comment by dreamcat4** [comment]
Well I just caught it out - mere seconds after starting it up. With `strace` and a pipe `|` to `tail -f`.

None of my own script code / while loops or anything like that. It played up right from the bat. I really hope that this `strace` output may reveal something to us and give any clues regarding the nature of this problem?

Below is a relevant condensed section, which is at the very end of the `strace` output.
PLUS a simple <----- annotation lines.

Full GIst: https://gist.github.com/dreamcat4/5081dc03919066c68164

```sh
getsockname(3, {sa_family=AF_LOCAL, NULL}, [2]) = 0
getpeername(3, {sa_family=AF_LOCAL, sun_path="/var/run/docker.sock"}, [23]) = 0
epoll_wait(4, {{EPOLLOUT, {u32=3727392840, u64=140002481373256}}}, 128, 0) = 1
read(3, 0xc208091000, 4096)             = -1 EAGAIN (Resource temporarily unavailable)
write(3, "GET /v1.17/events HTTP/1.1\r\nHost"..., 91) = 91
epoll_wait(4, {{EPOLLOUT, {u32=3727392840, u64=140002481373256}}}, 128, 0) = 1
epoll_wait(4, {{EPOLLIN|EPOLLOUT, {u32=3727392840, u64=140002481373256}}}, 128, -1) = 1
read(3, "HTTP/1.1 200 OK\r\nContent-Type: a"..., 4096) = 116
read(3, 0xc208091000, 4096)             = -1 EAGAIN (Resource temporarily unavailable)
epoll_wait(4, {}, 128, 0)               = 0
epoll_wait(4, {{EPOLLIN|EPOLLOUT, {u32=3727392840, u64=140002481373256}}}, 128, -1) = 1
futex(0x12eb440, FUTEX_WAKE, 1)         = 1
read(3, "90\r\n{\"status\":\"start\",\"id\":\"9aef"..., 4096) = 150
write(1, "2015-04-02T11:20:34.000000000+01"..., 36) = 36
write(1, "9aefebf40bddefeb00ea3f828f39a5d7"..., 66) = 66
write(1, "(from dreamcat4/tvheadend:latest"..., 34) = 34
write(1, "start\n", 6)                  = 6
read(3, 0xc208091000, 4096)             = -1 EAGAIN (Resource temporarily unavailable)
epoll_wait(4, {}, 128, 0)               = 0
epoll_wait(4, 
{{EPOLLIN|EPOLLOUT, {u32=3727392840, u64=140002481373256}}}, 128, -1) = 1
futex(0x12eb440, FUTEX_WAKE, 1)         = 1
futex(0x12eb3c0, FUTEX_WAKE, 1)         = 1
read(3, "8e\r\n{\"status\":\"die\",\"id\":\"9aefeb"..., 4096) = 148
write(1, "2015-04-02T11:20:42.000000000+01"..., 36) = 36
futex(0x12ec138, FUTEX_WAIT, 0, NULL

<------- here i run multiple container start-stop. And no new output from docker events process
 is 'seems' hung.


<------- here i kill the `docker events` process with `kill $pid` (SIGTERM)
<------- it flushes all of the missed events during process exit
Process 28503 detached
 <detached ...>
2015-04-02T11:20:34.000000000+01:00 9aefebf40bddefeb00ea3f828f39a5d7aace2ff6d52a1a7ecde61c4f3f24befb: (from dreamcat4/tvheadend:latest) start
2015-04-02T11:20:42.000000000+01:00 9aefebf40bddefeb00ea3f828f39a5d7aace2ff6d52a1a7ecde61c4f3f24befb: (from dreamcat4/tvheadend:latest) die
2015-04-02T11:20:42.000000000+01:00 9aefebf40bddefeb00ea3f828f39a5d7aace2ff6d52a1a7ecde61c4f3f24befb: (from dreamcat4/tvheadend:latest) stop
2015-04-02T11:20:55.000000000+01:00 9aefebf40bddefeb00ea3f828f39a5d7aace2ff6d52a1a7ecde61c4f3f24befb: (from dreamcat4/tvheadend:latest) start
2015-04-02T11:21:15.000000000+01:00 9aefebf40bddefeb00ea3f828f39a5d7aace2ff6d52a1a7ecde61c4f3f24befb: (from dreamcat4/tvheadend:latest) die
2015-04-02T11:21:15.000000000+01:00 9aefebf40bddefeb00ea3f828f39a5d7aace2ff6d52a1a7ecde61c4f3f24befb: (from dreamcat4/tvheadend:latest) stop
2015-04-02T11:21:42.000000000+01:00 9aefebf40bddefeb00ea3f828f39a5d7aace2ff6d52a1a7ecde61c4f3f24befb: (from dreamcat4/tvheadend:latest) start
id@emachines-e520:~/docker-images$ 
id@emachines-e520:~/docker-images$ 
id@emachines-e520:~/docker-images$ 
id@emachines-e520:~/docker-images$ 
```

**Comment by dreamcat4** [comment]
@cpuguy83 If you are interested: This may be a duplicate of #11207 - where you needed for more info than 'doesn't work'. It was a different guy who raised the other bug report. But I suspect it could the same issue he had. Please see the `strace` log above ^^. And please request me any further testing you need if you cannot reproduce it. Many thanks.

**Comment by dreamcat4** [comment]
FYI: Checked my docker version of `1.5.0` and it happens to be the same as the other guy `a8a31ef`.

```sh
 docker version
Client version: 1.5.0
Client API version: 1.17
Go version (client): go1.4.1
Git commit (client): a8a31ef
OS/Arch (client): linux/amd64
Server version: 1.5.0
Server API version: 1.17
Go version (server): go1.4.1
Git commit (server): a8a31ef
```

It would be great to move over and test behaviour on `1.6.x` also. I'm just a bit scared to, and also since we don't actually know what the problem is yet with `docker events` CLI command.

**Comment by Mashimiao** [comment]
Hi @dreamcat4 ,
I used the following commands:
for i in {1..1025};do docker run --rm ubuntu:latest echo 'hello'>/dev/null;sleep 300;done
and used `docker events` in another terminal.
It seems everything is OK.
==
$ docker version
Client version: 1.5.0-dev
Client API version: 1.19
Go version (client): go1.4.2
Git commit (client): 9ede3f9-dirty
OS/Arch (client): linux/amd64
Server version: 1.5.0-dev
Server API version: 1.19
Go version (server): go1.4.2
Git commit (server): 9ede3f9-dirty
OS/Arch (server): linux/amd64
[mashimiao@G08FNSTD121139 cpu]$ docker version
Client version: 1.5.0-dev
Client API version: 1.19
Go version (client): go1.4.2
Git commit (client): 9ede3f9-dirty
OS/Arch (client): linux/amd64
Server version: 1.5.0-dev
Server API version: 1.19
Go version (server): go1.4.2
Git commit (server): 9ede3f9-dirty
OS/Arch (server): linux/amd64

**Comment by dreamcat4** [comment]
@Mashimiao  Thanks but there seem a few differences with your report:

* You didn't pipe the output of `docker events` cmd in to `tail -f`.
* It is a slightly different docker version. We see only on git commit `a8a31ef` to our current testing.
* It is slightly newer go version (ours was 1.4.1).
* It is slightly newer docker api version (ours was 1.17.


If you have any further opportunity, then please re-run your success case test again, but this time with `strace` cmd in front of it. and pipe to `tail -f` too. e.g. like this:

```sh
strace docker events | tail -f
```
Many thanks.

@Mashimiao  I will eventually upgrade myself, to re-test. However I was waiting until `1.6.0` comes released. But now I would like to re-test with your version. How is your 1.5.0 dev version currently available? In docker APT repository?

**Comment by Mashimiao** [comment]
I just clone docker's source code, then build and install by myself.
I will retest with `strace docker events | tail -f`

**Comment by dreamcat4** [comment]
@Mashimiao Thanks. I can't be bothered to recompile myself, but will jump further forwards pas that one, and upgrade to `1.6.0-RC5` soon-ish. As that is available in APT testing repo.

**Comment by dreamcat4** [comment]
Hello again!

Sorry to say: but I upgraded my docker and re-tested. It reproduced again this same exact issue on latest official `docker-1.6.0-RC5`. Problem appeared immediately, when running like this: `strace docker events | tail -f`. The `strace` output looks to be the same:

https://gist.github.com/dreamcat4/537877f7b5a088797b4f

Bug did not happen with just `docker events` cmd only (e.g. in interactive terminal). You must be sending the output through a UNIX pipe `|` to be received by some other command.

@Mashimiao You may not need to `sleep` of such long times to see this bug. Just start-stop your container a few times manually. While in a 2nd window listening with docker events to PIPE to `| tail -f`.

Please also provide your `docker info`. Many thanks.

```sh
root@emachines-e520:~# docker info
Containers: 7
Images: 170
Storage Driver: aufs
Root Dir: /var/lib/docker/aufs
Backing Filesystem: extfs
Dirs: 184
Dirperm1 Supported: true
Execution Driver: native-0.2
Kernel Version: 3.16.0-33-generic
Operating System: Ubuntu 14.10
CPUs: 2
Total Memory: 3.792 GiB
Name: emachines-e520
ID: NG2B:QLY4:W2SH:6FQF:FYGU:TVLX:3IQL:CQ4S:N5DL:CTND:52ZD:GCDH
```

**Comment by dreamcat4** [comment]
As previously. We have `epoll_wait` it hangs there. Some thread blocking or something like that. Presumably Go has `epoll` wrapped in it's own API library.

```sh
read(3, 0xc2080cd000, 4096) = -1 EAGAIN (Resource temporarily unavailable)
epoll_wait(4, {}, 128, 0) = 0
futex(0x134e1b8, FUTEX_WAIT, 0, NULL
 
<--------- here are multiple container start-stop. No new output from docker events process. It seems 'hung'.
 
 
<--------- here i keill the `docker events` process with `kill $pid` (SIGTERM)
<--------- it flushed some or all of the missed events during it's process cleanup handler) = 0
epoll_wait(4, {}, 128, 0) = 0
futex(0x134d480, FUTEX_WAKE, 1) = 1
epoll_wait(4, {}, 128, 0) = 0
futex(0x134e1b8, FUTEX_WAIT, 0, NULLProcess 5197 detached
<detached ...> 
```

**Comment by dreamcat4** [comment]
I am guessing in the docker source code it is at here:

https://github.com/docker/docker/blob/v1.6.0-rc5/api/client/commands.go#L1877

Sorry I don't understand that code where it goes into the function `cli.stream`. It is under the surface. Probably relying upon whatever standard GOlang I/O libraries to achieve that.

In addition, when I add a unix pipe on command line (to the end of that chain). Then somehow we end up in trouble and something breaks. As per the `strace` output. Where the `futex(0x134e1b8, FUTEX_WAIT, 0, NULL  `. Does not finish printing that `strace` message.

I guess this is the relevant part of (new) strace output:

```sh
socket(PF_LOCAL, SOCK_STREAM|SOCK_CLOEXEC|SOCK_NONBLOCK, 0) = 3
setsockopt(3, SOL_SOCKET, SO_BROADCAST, [1], 4) = 0
connect(3, {sa_family=AF_LOCAL, sun_path="/var/run/docker.sock"}, 23) = 0
epoll_create1(EPOLL_CLOEXEC) = 4
epoll_ctl(4, EPOLL_CTL_ADD, 3, {EPOLLIN|EPOLLOUT|EPOLLRDHUP|EPOLLET, {u32=1040633704, u64=140442176245608}}) = 0
getsockname(3, {sa_family=AF_LOCAL, NULL}, [2]) = 0
getpeername(3, {sa_family=AF_LOCAL, sun_path="/var/run/docker.sock"}, [23]) = 0
read(3, 0xc2080cd000, 4096) = -1 EAGAIN (Resource temporarily unavailable)
write(3, "GET /v1.18/events HTTP/1.1\r\nHost"..., 95) = 95
epoll_wait(4, {{EPOLLIN|EPOLLOUT, {u32=1040633704, u64=140442176245608}}}, 128, 0) = 1
read(3, "HTTP/1.1 200 OK\r\nContent-Type: a"..., 4096) = 116
read(3, 0xc2080cd000, 4096) = -1 EAGAIN (Resource temporarily unavailable)
epoll_wait(4, {}, 128, 0) = 0
epoll_wait(4, {{EPOLLIN|EPOLLOUT, {u32=1040633704, u64=140442176245608}}}, 128, -1) = 1
futex(0x134d480, FUTEX_WAKE, 1) = 1
read(3, "96\r\n{\"status\":\"start\",\"id\":\"fbc9"..., 4096) = 156
write(1, "2015-04-09T11:12:07.000000000+01"..., 36) = 36
write(1, "fbc922fbb30bb7c8eaf84cb87daec1a8"..., 66) = 66
write(1, "(from dreamcat4/tvheadend.debug:"..., 40) = 40
write(1, "start\n", 6) = 6
futex(0x134d480, FUTEX_WAKE, 1) = 1
futex(0x134d400, FUTEX_WAKE, 1) = 1
read(3, 0xc2080cd000, 4096) = -1 EAGAIN (Resource temporarily unavailable)
epoll_wait(4, {}, 128, 0) = 0
futex(0x134e1b8, FUTEX_WAIT, 0, NULL
```

**Comment by dreamcat4** [comment]
@vieux @shin- If you can please tell us where this line goes to?

```golang
if err := cli.stream("GET", "/events?"+v.Encode(), nil, cli.out, nil); err != nil
```

https://github.com/docker/docker/blob/v1.6.0-rc5/api/client/commands.go#L1877

**Comment by LK4D4** [comment]
I can reproduce same behavior with `tail -f logifle | tail -f`. I think this is how `tail -f` working.

**Comment by dreamcat4** [comment]
It's not just the `tail -f` command. I write the output to a unix socket with `mkfifo file.sock`, then `docker events > file.sock`. And it still gets the same bug -

https://github.com/dreamcat4/docker-images/blob/master/pipework/entrypoint.sh#L368

**Comment by dreamcat4** [comment]
I still assume that the buffering is being left to some go IO library underneath, sorry I cannot follow to where in docker's code.

Perhaps an IO reader / writer object may not be flushed properly for some reason. The eventing for outputting (IO writer) may occur in same thread that is being used to gobble (IO read) the input from docker daemon response. Or might the IO object require a different configuration (different options or flags) for non-interactive (no-tty) situation. A 3rd possibility may be some genuine error in the go library underneath. Or error in the docker handler object which wraps around these IO calls. There is nothing really to see directly in the foremost `cli.go` / `events.go` file.

We need to see deeper in docker source code. Where is it happening? Please help if you can.

**Comment by LK4D4** [comment]
@dreamcat4 client-side is using `cli.streamBody` where it just doing `io.Copy` from `io.ReadCloser` which is basically response body to `cli.out`, which is `os.Stdout`. On daemon side it uses `ResponseWriter` wrapped in `utils.NewFlusher` and `json.NewEncoder`. Because of `utils.NewFlusher` I think that from daemon side all flushed okay.
Flush problems can be with `os.Stdout`, because it is basically `*os.File`.

**Comment by LK4D4** [comment]
@dreamcat4 Also you need to read other side of pipe to write to pipe. It works for me.

**Comment by dreamcat4** [comment]
>  I think that from daemon side all flushed okay.

Yes I agree. And that matches the kill behaviour. Because then the missing messsages appear.

> Flush problems can be with os.Stdout, because it is basically *os.File.

@LKD4

* If it works for you, please provide an example here so I can try to reproduce that.

* Please provide source code references on client side? E.g. where the `cli.streambody`, and `io.ReadCloser` get invoked?

**Comment by LK4D4** [comment]
@dreamcat4 `cli.streamBody` is in `api/client/utils.go`. You should get go-to-definition plugin for go - it'll simplify your life a lot.
I took reader example from http://www.linuxjournal.com/content/using-named-pipes-fifos-bash
And in other terminal run `docker events > /tmp/testpipe`

**Comment by dreamcat4** [comment]
Ok. So if way you called `read` was like this:

```sh
while true
do
    if read line <$pipe; then
        if [[ "$line" == 'quit' ]]; then
            break
        fi
        echo $line
    fi
done
```

Wheras the way I call `read` is like this:

```sh
    while read event_line; do
        container_id="$(echo -e " $event_line" | grep -v "from $_pipework_image_name" | tr -s ' ' | cut -d ' ' -f3)"
        [ "$container_id" ] && _process_container ${container_id%:};
    done < $_docker_events_fifo &
```

https://github.com/dreamcat4/docker-images/blob/master/pipework/entrypoint.sh#L368

Then you would have me do this instead:

```sh
while true; do
    if read event_line < $_docker_events_fifo; then
        container_id="$(echo -e " $event_line" | grep -v "from $_pipework_image_name" | tr -s ' ' | cut -d ' ' -f3)"
        [ "$container_id" ] && _process_container ${container_id%:};
    fi
done &
```

I can try that tomorrow. Sorry, it is late here now.

**Comment by Mashimiao** [comment]
Hi @dreamcat4 ,
I used `docker events` in terminal A, `strace docker events` in terminal B and `strace docker events | tail -f` in terminal C with the above testcase in parallel. But lessen sleep time to 3s.
Well, it's very interesting.
Command in terminal A and B will work very well.
But command in terminal C will hung after running for a while.
And then, in terminal D, I used `docker event > /tmp/event_log &`  and `tail -f /tmp/event_log`, run the testcase again.
Commands in terminal D also work very well.

So I think the problem occurs because you used pipeline directly to read output of `strace docker events`.

**Comment by dreamcat4** [comment]
@LK4D4 This is weird. Not expected. Because today when I try either socket ways (mine or yours) in an interactive terminal - they seem to both work.

Nevertheless I have made your suggested change inside my docker image. Because that is a not interactive - it is shell script. So we shall see (after a while more) if it makes any difference.

@Mashimiao Thanks for the tip. About writing to log file. I am not actually using `tail -f` command in real code. That was just a simple test case to reproduce the problem. So you guys did not need some special shell script.

Sorry I cannot reproduce (my ways). But here was the test I ran:

```sh
    # Setup Test 1 - In window 1 - dreamcat's way
    _docker_events_fifo1="/tmp/docker-events.sock1"
    mkfifo --mode 0600 $_docker_events_fifo1

    _docker_events_fifo1="/tmp/docker-events.sock1"
    while read event_line; do
        echo $event_line
    done < $_docker_events_fifo1 &

    _docker_events_fifo1="/tmp/docker-events.sock1"
    docker events > $_docker_events_fifo1 &
    fg 1

    # Setup Test 2 - In Window 2 - by KD4D4 - correct way
    _docker_events_fifo2="/tmp/docker-events.sock2"
    mkfifo --mode 0600 $_docker_events_fifo2

    _docker_events_fifo2="/tmp/docker-events.sock2"
    while true; do
        if read event_line < $_docker_events_fifo2; then
            echo $event_line
        fi
    done &

    _docker_events_fifo2="/tmp/docker-events.sock2"
    docker events > $_docker_events_fifo2 &
    fg 1

    # Window 3 - Start / Stop some container a few times
    my_container=tvh.debug
    docker start $my_container
    docker stop  $my_container
    docker start $my_container
    docker stop  $my_container
    docker start $my_container
    docker stop  $my_container
    docker start $my_container
    docker stop  $my_container

```

**Comment by dreamcat4** [comment]
@LK4D4 OK:

With that small change to the `while read` line. Put inside of my non-interactive shell script (the real code), I have tried catching the error again. The testing was not extremely thorough. However no issue to be seen so far. Continued to work for multiple start / stop events.

To my memory, the bug should have occurred after the first newline `\n` character. So I am happy for the time being with this resolution and push this change up to my official docker image on dockerhub. Many thanks. I just hope that it continues to work without coming back again.

Still `docker events | tail -f` does not work. However not concerned about that myself. Thanks again. Appreciate the help.

**Comment by LK4D4** [comment]
I think we can close this, because `tail -f file | tail -f` not working as well.
Feel free to open new issue if you find something interesting.

**Comment by dreamcat4** [comment]
Nope. Didn't work. Problem still exists.

**Comment by dreamcat4** [comment]
@Mashimiao Using your log file in /tmp - that works in conjunction with the polling method that was pointed to by @LK4D4's reference `read` example;

```sh
    _docker_events_log=/tmp/docker-events.log
    while true; do
        if read event_line < $_docker_events_log; then
            echo $event_line
        fi
    done &
```

However the constant polling is **very** bad and cannot be used. Docker daemon writes a log line for every loop, each call to `read`. That causes all available HDD space to be quickly consumed by the `-json.log` file in `containers/` subfolder. And also the Daemon process ended up consuming approximately `2.6GB` of ram.

So no choice but to be going back to relying upon `stdio` and `EOF` like this:

```sh
    while read event_line; do
        container_id="$(echo -e " $event_line" | grep -v "from $_pipework_image_name" | tr -s ' ' | cut -d ' ' -f3)"
        [ "$container_id" ] && _process_container ${container_id%:};
    done < $_docker_events_log &
```

And hey presto - back to being broken again. In same manner / symptoms as per the original bug report - no flush. Thus no new lines fed to `read`.

Problem not solved. In these circumstances all the fingers are pointing to the underlying Go libraries. Since not any Docker code to look at. But, @LK4D4 do you feel maybe it's justified to write a test case to reproduce this? Surely I can't be the only person experiencing this problem?

Actually I must say it is not at all unreasonable both your suggested solutions. Just except ^^ for the `if` polling eating all {hdd, ram} resources on the local machine.

**Comment by dreamcat4** [comment]
OK. I think maybe I should try switching over to using the Docker API `GET /events` endpoint request instead of the CLI tool.

In fact, `curl` is being used in such a manner shown by this ticket:

https://github.com/docker/docker/issues/12216

I just need to figure out how to translate these "curl" API requests onto the `docker.sock`. Hmm, never done that before.

**Comment by dreamcat4** [comment]
This seems better ways:

http://stackoverflow.com/a/11291927/287510

Just testing it now.

**Comment by dreamcat4** [comment]
Problem found: Don't use fifo (unix socket) - that causes the hangs. Since I tried both using curl / netcat and the CLI command. Both broken with socket.

Polling problem has been solved for reading docker log.

```sh
_cleanup ()
{
    [ "$_while_read_pid" ]     && kill  $_while_read_pid
    [ "$_docker_events_pid" ]  && kill  $_docker_events_pid
    [ "$_tail_f_pid" ]         && kill  $_tail_f_pid
    [ "$_docker_events_log" ]  && rm -f $_docker_events_log
    exit 0
}
trap _cleanup TERM INT QUIT HUP

    # Create docker events log
    _docker_events_log="/tmp/docker-events.log"
    rm -f $_docker_events_log
    touch $_docker_events_log
    chmod 0600 $_docker_events_log

    # http://stackoverflow.com/questions/1652680/how-to-get-the-pid-of-a-process-that-is-piped-to-another-process-in-bash
    tail_f_pid_file="$(mktemp -u --suffix=.pid /var/run/tail_f.XXX)"
    ( tail -f $_docker_events_log & echo $! >&3 ) 3>$tail_f_pid_file | \
    while true
    do
        read event_line
        echo event_line=$event_line
    done &
    _while_read_pid=$!
    _tail_f_pid=$(cat $tail_f_pid_file) && rm -f $tail_f_pid_file

    # Start to listen for new container start events and write them to the events log
    docker events $_pe_opts --filter='event=start' --filter='event=die' \
        $_pipework_daemon_event_opts > $_docker_events_log &
    _docker_events_pid=$!

    # Wait until 'docker events' command is killed by 'trap _cleanup ...'
    wait $_docker_events_pid
    _cleanup;
```
